File: src/pci/pci.c
#include "pci.h"
#include "mmio/mmio.h"
#include "uart/uart.h"
#include "utils/utils.h"

static struct pci_device pci_devices[MAX_PCI_DEVICES];
static uint16_t pci_device_count = 0;

static uintptr_t mmio32_next = PCI_MMIO32_PHYS;
static uintptr_t mmio64_next = PCI_MMIO64_PHYS;

static uintptr_t alloc_mmio32(uint32_t size) {
  /* Align up to the BAR's natural alignment */
  uintptr_t mask = (uintptr_t)size - 1;
  mmio32_next = (mmio32_next + mask) & ~mask;

  if (mmio32_next & (size - 1)) {
    mmio32_next = (mmio32_next + size) & ~(size - 1);
  }

  uintptr_t addr = mmio32_next;
  mmio32_next += size;
  return addr;
}

static uintptr_t alloc_mmio64(uint64_t size) {
  /* Align up to the BAR's natural alignment */
  uintptr_t mask = (uintptr_t)size - 1;
  mmio64_next = (mmio64_next + mask) & ~mask;

  if (mmio64_next & (size - 1)) {
    mmio64_next = (mmio64_next + size) & ~(size - 1);
  }

  uintptr_t addr = mmio64_next;
  mmio64_next += size;
  return addr;
}

// ECAM addresses passed to mmio_read/write are physical.
// The MMIO layer adds KERNEL_VA_OFFSET (after mmio_switch_to_upper)
// to access them through TTBR1 upper half.

static uintptr_t pci_make_ecam_addr(uint16_t bus, uint8_t slot, uint8_t func,
                                    uint16_t offset) {
  return PCI_ECAM_PHYS | ((uintptr_t)bus << 20) | ((uintptr_t)slot << 15) |
         ((uintptr_t)func << 12) | (uintptr_t)offset;
}

uint32_t pci_config_read32(uint16_t bus, uint8_t slot, uint8_t func,
                           uint16_t offset) {
  return mmio_read32(pci_make_ecam_addr(bus, slot, func, offset));
}

uint16_t pci_config_read16(uint16_t bus, uint8_t slot, uint8_t func,
                           uint16_t offset) {
  return mmio_read16(pci_make_ecam_addr(bus, slot, func, offset));
}

uint8_t pci_config_read8(uint16_t bus, uint8_t slot, uint8_t func,
                         uint16_t offset) {
  return mmio_read8(pci_make_ecam_addr(bus, slot, func, offset));
}

void pci_config_write32(uint16_t bus, uint8_t slot, uint8_t func,
                        uint16_t offset, uint32_t val) {
  mmio_write32(pci_make_ecam_addr(bus, slot, func, offset), val);
}

void pci_config_write16(uint16_t bus, uint8_t slot, uint8_t func,
                        uint16_t offset, uint16_t val) {
  mmio_write16(pci_make_ecam_addr(bus, slot, func, offset), val);
}

void pci_config_write8(uint16_t bus, uint8_t slot, uint8_t func,
                       uint16_t offset, uint8_t val) {
  mmio_write8(pci_make_ecam_addr(bus, slot, func, offset), val);
}

static void pci_log_device_found(uint16_t bus, uint8_t slot, uint8_t func,
                                 uint16_t vendor_id, uint16_t device_id) {
  uart_printf("[PCI] Device found at %d:%d.%d | VendorID: %x, DeviceID: %x\n",
              bus, slot, func, vendor_id, device_id);
}

void pci_enumerate_bus() {
  uart_println("[PCI] Enumerating PCI Devices");

  for (uint16_t bus = 0; bus < MAX_PCI_BUS; bus++) {
    for (uint8_t slot = 0; slot < MAX_PCI_SLOT; slot++) {
      for (uint8_t func = 0; func < MAX_PCI_FUNC; func++) {
        uint16_t vendor_id = pci_config_read16(bus, slot, func, PCI_VENDOR_ID);

        if (vendor_id == 0xFFFF) {
          continue;
        }

        uint16_t device_id = pci_config_read16(bus, slot, func, PCI_DEVICE_ID);
        pci_log_device_found(bus, slot, func, vendor_id, device_id);

        if (pci_device_count >= MAX_PCI_DEVICES) {
          uart_errorln("[PCI] Max PCI devices limit reached");
          return;
        }

        pci_devices[pci_device_count] = (struct pci_device){
            .bus = bus,
            .slot = slot,
            .func = func,
            .vendor_id = vendor_id,
            .device_id = device_id,
        };
        pci_device_count++;
      }
    }
  }
}

int pci_find_device(uint16_t vendor_id, uint16_t device_id,
                    struct pci_device *pci_device) {
  for (uint16_t i = 0; i < pci_device_count; i++) {
    if (pci_devices[i].vendor_id == vendor_id &&
        pci_devices[i].device_id == device_id) {
      *pci_device = pci_devices[i];
      return ESUCCESS;
    }
  }
  return EERROR;
}

uint8_t pci_get_header_type(struct pci_device *dev) {
  uint8_t header_type =
      pci_config_read8(dev->bus, dev->slot, dev->func, PCI_HEADER_TYPE);
  return header_type;
}

static uint32_t pci_get_bar_size(uint8_t bus, uint8_t slot, uint8_t func,
                                 uint16_t offset) {
  uint32_t original = pci_config_read32(bus, slot, func, offset);
  pci_config_write32(bus, slot, func, offset, 0xFFFFFFFF);
  uint32_t size_mask = pci_config_read32(bus, slot, func, offset);
  pci_config_write32(bus, slot, func, offset, original);

  // mask lower 4 bits (control bits)
  size_mask &= ~0xF;
  if (size_mask == 0) {
    // unimplemented BAR
    return 0;
  }
  uint32_t size = ~size_mask + 1;

  return size;
}

// For 64-bit BARs, probe both BAR[i] and BAR[i+1] to get the full size
static uint64_t pci_get_bar_size64(uint8_t bus, uint8_t slot, uint8_t func,
                                   uint16_t offset_lo, uint16_t offset_hi) {
  // Save originals
  uint32_t orig_lo = pci_config_read32(bus, slot, func, offset_lo);
  uint32_t orig_hi = pci_config_read32(bus, slot, func, offset_hi);

  // Write all-1s to both halves
  pci_config_write32(bus, slot, func, offset_lo, 0xFFFFFFFF);
  pci_config_write32(bus, slot, func, offset_hi, 0xFFFFFFFF);

  // Read back
  uint32_t mask_lo = pci_config_read32(bus, slot, func, offset_lo);
  uint32_t mask_hi = pci_config_read32(bus, slot, func, offset_hi);

  // Restore originals
  pci_config_write32(bus, slot, func, offset_lo, orig_lo);
  pci_config_write32(bus, slot, func, offset_hi, orig_hi);

  // Mask lower 4 control bits of the low word
  uint64_t mask = ((uint64_t)mask_hi << 32) | (mask_lo & ~0xFUL);
  uint64_t size = ~mask + 1;

  return size;
}

void pci_assign_bars(struct pci_device *dev) {
  uart_println("[PCI] Assigning BARs");

  uint8_t b = dev->bus;
  uint8_t d = dev->slot;
  uint8_t f = dev->func;

  for (uint8_t i = 0; i < 6; i++) {
    uint32_t bar_offset = PCI_BAR0 + i * 4;
    uint32_t bar = pci_config_read32(b, d, f, bar_offset);

    /* https://wiki.osdev.org/PCI */
    /* I/O Space BAR Layout */
    if (bar & 0x01) {
      uart_errorln("[PCI] IO BAR Type, Ignoring");
      continue;
    }

    /* Memory Space BAR Layout */
    uint8_t type = (bar >> 1) & 0x03;

    if (type == 0x00) {
      /* 32-bit BAR */
      uint32_t size = pci_get_bar_size(b, d, f, bar_offset);

      if (size == 0 || size == 0xFFFFFFFF) {
        dev->bar_addr[i] = 0;
        continue;
      }

      uart_println("[PCI][32 Bit Memory Space]");
      uart_puts(" BAR");
      uart_putdec(i);
      uart_puts(" has size: ");
      uart_puthex(size);
      uart_println("");
      uintptr_t addr = alloc_mmio32(size);

      pci_config_write32(b, d, f, bar_offset, (uint32_t)addr);
      dev->bar_addr[i] = addr;
    } else if (type == 0x02) {
      /* 64-bit BAR — probe both halves for correct size */
      uint32_t bar_offset_hi = PCI_BAR0 + (i + 1) * 4;
      uint64_t size64 = pci_get_bar_size64(b, d, f, bar_offset, bar_offset_hi);

      if (size64 == 0) {
        dev->bar_addr[i] = 0;
        i++; // skip upper BAR
        continue;
      }

      uart_println("[PCI][64 Bit Memory Space]");
      uart_puts(" BAR");
      uart_putdec(i);
      uart_puts(" has size: ");
      uart_puthex(size64);
      uart_println("");
      uintptr_t addr = alloc_mmio64(size64);
      // lower 32 bit
      pci_config_write32(b, d, f, bar_offset, (uint32_t)(addr & 0xFFFFFFFF));
      // upper 32 bit
      pci_config_write32(b, d, f, bar_offset + 4, (uint32_t)(addr >> 32));
      dev->bar_addr[i] = addr;
      i++;
    } else {
      uart_errorln("[PCI][Memory Space Type] Huh ?");
    }
  }

  uart_println("[PCI] BARs Assigned");
}

void pci_enable_device(struct pci_device *dev) {
  uart_println("[PCI] Enabling device");

  /* https://wiki.osdev.org/PCI#Command_Register */
  uint16_t cmd = pci_config_read16(dev->bus, dev->slot, dev->func, PCI_COMMAND);
  cmd |= (1 << 1); // Memory Space Enable
  cmd |= (1 << 2); // Bus Master Enable (DMA)

  pci_config_write16(dev->bus, dev->slot, dev->func, PCI_COMMAND, cmd);
  uart_println("[PCI] Device Enabled");
}

File: src/pci.rs
//! PCI Express (ECAM) — brute-force enumeration, BAR assignment, enable.
//!
//! Port of `src/pci/pci.c` for the QEMU virt machine.

use crate::kprintln;
use crate::mmio;
use crate::sync::Racy;
use crate::uart;

pub const PCI_ECAM_PHYS: usize = 0x40_1000_0000;
pub const PCI_MMIO32_PHYS: u64 = 0x1000_0000;
pub const PCI_MMIO32_LIMIT: u64 = 0x3EFE_FFFF;
pub const PCI_MMIO64_PHYS: u64 = 0x80_0000_0000;
pub const PCI_MMIO64_LIMIT: u64 = 0xFF_FFFF_FFFF;

const MAX_PCI_DEVICES: usize = 16;
const MAX_PCI_BUS: u16 = 256;
const MAX_PCI_SLOT: u8 = 32;
const MAX_PCI_FUNC: u8 = 8;

pub const PCI_VENDOR_ID: u16 = 0x00;
pub const PCI_DEVICE_ID: u16 = 0x02;
pub const PCI_COMMAND: u16 = 0x04;
pub const PCI_STATUS: u16 = 0x06;
pub const PCI_HEADER_TYPE: u16 = 0x0E;
pub const PCI_BAR0: u16 = 0x10;
pub const PCI_CAP_PTR: u16 = 0x34;
pub const PCI_INTERRUPT_PIN: u16 = 0x3D;
pub const PCI_ENDPOINT_DEV_TYPE: u8 = 0x00;

#[derive(Clone, Copy)]
pub struct PciDevice {
    pub bus: u8,
    pub slot: u8,
    pub func: u8,
    pub vendor_id: u16,
    pub device_id: u16,
    pub bar_addr: [u64; 6],
}

impl PciDevice {
    const fn empty() -> Self {
        Self {
            bus: 0,
            slot: 0,
            func: 0,
            vendor_id: 0,
            device_id: 0,
            bar_addr: [0; 6],
        }
    }
}

struct PciState {
    devices: [PciDevice; MAX_PCI_DEVICES],
    count: u16,
    mmio32_next: u64,
    mmio64_next: u64,
}

static PCI: Racy<PciState> = Racy::new(PciState {
    devices: [PciDevice::empty(); MAX_PCI_DEVICES],
    count: 0,
    mmio32_next: PCI_MMIO32_PHYS,
    mmio64_next: PCI_MMIO64_PHYS,
});

fn ecam_addr(bus: u16, slot: u8, func: u8, offset: u16) -> usize {
    PCI_ECAM_PHYS
        | ((bus as usize) << 20)
        | ((slot as usize) << 15)
        | ((func as usize) << 12)
        | offset as usize
}

pub fn config_read32(bus: u16, slot: u8, func: u8, off: u16) -> u32 {
    mmio::read32(ecam_addr(bus, slot, func, off))
}
pub fn config_read16(bus: u16, slot: u8, func: u8, off: u16) -> u16 {
    mmio::read16(ecam_addr(bus, slot, func, off))
}
pub fn config_read8(bus: u16, slot: u8, func: u8, off: u16) -> u8 {
    mmio::read8(ecam_addr(bus, slot, func, off))
}
pub fn config_write32(bus: u16, slot: u8, func: u8, off: u16, v: u32) {
    mmio::write32(ecam_addr(bus, slot, func, off), v)
}
pub fn config_write16(bus: u16, slot: u8, func: u8, off: u16, v: u16) {
    mmio::write16(ecam_addr(bus, slot, func, off), v)
}

pub fn enumerate_bus() {
    uart::println("[PCI] Enumerating PCI Devices");
    let p = unsafe { PCI.get() };
    for bus in 0..MAX_PCI_BUS {
        for slot in 0..MAX_PCI_SLOT {
            for func in 0..MAX_PCI_FUNC {
                let vendor_id = config_read16(bus, slot, func, PCI_VENDOR_ID);
                if vendor_id == 0xFFFF {
                    continue;
                }
                let device_id = config_read16(bus, slot, func, PCI_DEVICE_ID);
                kprintln!(
                    "[PCI] Device at {}:{}.{} | Vendor: {:#x} Device: {:#x}",
                    bus,
                    slot,
                    func,
                    vendor_id,
                    device_id
                );
                if p.count as usize >= MAX_PCI_DEVICES {
                    uart::errorln("[PCI] Max PCI devices reached");
                    return;
                }
                p.devices[p.count as usize] = PciDevice {
                    bus: bus as u8,
                    slot,
                    func,
                    vendor_id,
                    device_id,
                    bar_addr: [0; 6],
                };
                p.count += 1;
            }
        }
    }
}

/// Find a device by vendor/device id (returns a copy).
pub fn find_device(vendor_id: u16, device_id: u16) -> Option<PciDevice> {
    let p = unsafe { PCI.get() };
    for i in 0..p.count as usize {
        if p.devices[i].vendor_id == vendor_id && p.devices[i].device_id == device_id {
            return Some(p.devices[i]);
        }
    }
    None
}

pub fn header_type(dev: &PciDevice) -> u8 {
    config_read8(dev.bus as u16, dev.slot, dev.func, PCI_HEADER_TYPE)
}

fn bar_size32(bus: u16, slot: u8, func: u8, off: u16) -> u32 {
    let original = config_read32(bus, slot, func, off);
    config_write32(bus, slot, func, off, 0xFFFF_FFFF);
    let mut mask = config_read32(bus, slot, func, off);
    config_write32(bus, slot, func, off, original);
    mask &= !0xF;
    if mask == 0 {
        return 0;
    }
    (!mask).wrapping_add(1)
}

fn bar_size64(bus: u16, slot: u8, func: u8, lo: u16, hi: u16) -> u64 {
    let orig_lo = config_read32(bus, slot, func, lo);
    let orig_hi = config_read32(bus, slot, func, hi);
    config_write32(bus, slot, func, lo, 0xFFFF_FFFF);
    config_write32(bus, slot, func, hi, 0xFFFF_FFFF);
    let mask_lo = config_read32(bus, slot, func, lo);
    let mask_hi = config_read32(bus, slot, func, hi);
    config_write32(bus, slot, func, lo, orig_lo);
    config_write32(bus, slot, func, hi, orig_hi);
    let mask = ((mask_hi as u64) << 32) | (mask_lo & !0xF) as u64;
    (!mask).wrapping_add(1)
}

fn alloc_mmio32(size: u32) -> u64 {
    let p = unsafe { PCI.get() };
    let mask = (size as u64) - 1;
    p.mmio32_next = (p.mmio32_next + mask) & !mask;
    if p.mmio32_next + size as u64 - 1 > PCI_MMIO32_LIMIT {
        uart::errorln("[PCI] 32-bit MMIO window exhausted");
        return 0;
    }
    let addr = p.mmio32_next;
    p.mmio32_next += size as u64;
    addr
}

fn alloc_mmio64(size: u64) -> u64 {
    let p = unsafe { PCI.get() };
    let mask = size - 1;
    p.mmio64_next = (p.mmio64_next + mask) & !mask;
    if p.mmio64_next + size - 1 > PCI_MMIO64_LIMIT {
        uart::errorln("[PCI] 64-bit MMIO window exhausted");
        return 0;
    }
    let addr = p.mmio64_next;
    p.mmio64_next += size;
    addr
}

/// Assign MMIO BARs and write them into the device's `bar_addr`.
pub fn assign_bars(dev: &mut PciDevice) {
    uart::println("[PCI] Assigning BARs");
    let (b, d, f) = (dev.bus as u16, dev.slot, dev.func);
    let mut i = 0u8;
    while i < 6 {
        let off = PCI_BAR0 + (i as u16) * 4;
        let bar = config_read32(b, d, f, off);
        if bar & 0x01 != 0 {
            i += 1;
            continue; // I/O BAR
        }
        let kind = (bar >> 1) & 0x03;
        if kind == 0x00 {
            let size = bar_size32(b, d, f, off);
            if size == 0 || size == 0xFFFF_FFFF {
                dev.bar_addr[i as usize] = 0;
                i += 1;
                continue;
            }
            let addr = alloc_mmio32(size);
            config_write32(b, d, f, off, addr as u32);
            dev.bar_addr[i as usize] = addr;
            kprintln!("[PCI]  BAR{} 32-bit size={:#x} -> {:#x}", i, size, addr);
        } else if kind == 0x02 {
            if i + 1 >= 6 {
                dev.bar_addr[i as usize] = 0;
                i += 1;
                continue;
            }
            let off_hi = PCI_BAR0 + ((i + 1) as u16) * 4;
            let size = bar_size64(b, d, f, off, off_hi);
            if size == 0 {
                dev.bar_addr[i as usize] = 0;
                i += 2;
                continue;
            }
            let addr = alloc_mmio64(size);
            config_write32(b, d, f, off, (addr & 0xFFFF_FFFF) as u32);
            config_write32(b, d, f, off_hi, (addr >> 32) as u32);
            dev.bar_addr[i as usize] = addr;
            kprintln!("[PCI]  BAR{} 64-bit size={:#x} -> {:#x}", i, size, addr);
            i += 2;
            continue;
        }
        i += 1;
    }
    uart::println("[PCI] BARs Assigned");
}

pub fn enable_device(dev: &PciDevice) {
    let mut cmd = config_read16(dev.bus as u16, dev.slot, dev.func, PCI_COMMAND);
    cmd |= 1 << 1; // Memory Space Enable
    cmd |= 1 << 2; // Bus Master Enable (DMA)
    config_write16(dev.bus as u16, dev.slot, dev.func, PCI_COMMAND, cmd);
    uart::println("[PCI] Device Enabled");
}

/// GIC INTID for a PCI device's legacy INTx pin on the QEMU virt machine.
/// From the board's PCIe interrupt-map: SPI = 3 + ((slot + pin-1) % 4),
/// INTID = 32 + SPI. Returns 0 if the device has no interrupt pin.
pub fn device_intid(dev: &PciDevice) -> u32 {
    let pin = config_read8(dev.bus as u16, dev.slot, dev.func, PCI_INTERRUPT_PIN);
    if pin == 0 {
        return 0;
    }
    let swizzle = (dev.slot as u32 + (pin as u32 - 1)) % 4;
    35 + swizzle
}
